### ISSUE TITLE
nimsuggest devours CPU usage

### ISSUE THREAD

**Issue by s0kil** [nim-lang/Nim#11679]
`nimsuggest` is consuming too much of CPU, fans have not stopped since I installed it via choosenim. In fact it uses more CPU than any other process.
![Screenshot from 2019-07-06 21-03-42](https://user-images.githubusercontent.com/24403355/60762594-16a17600-a032-11e9-8ffd-36064be8c46c.png)

```
Nim Compiler Version 0.20.0 [Linux: amd64]
Compiled at 2019-06-06
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: e7471cebae2a404f3e4239f199f5a0c422484aac
active boot switches: -d:release
```

**Comment by dom96** [comment]
Are you using VS Code? If so that's the fault of the VS code Nim extension. It spawns Nimsuggest far too aggressively.

**Comment by sinkingsugar** [comment]
I don't think VS Code has anything to do, I mean it might be aggressive but here we are talking of full CPU spinning.. there has to be something wrong.

**Comment by mratsim** [comment]
There are multiple issues with nimsuggest.

1. Like OP said, in some cases, nimsuggest just keep spinning, and eat all CPU and need to be killall regularly. I've notice it happened often while I'm in a middle of a macro. Worse, on Windows that will lock the file and causes issues with git.

2. Like dom96 the VScode Nim extension is very aggressive because it spawns one nimsuggest per file instead of per module which means that combined with 1 multiple instances of nimsuggest can be stuck at 100%

**Comment by cooldome** [comment]
The only way to make nimsuggest work for me in the large project is to spawn nimsuggest as aggressively as possible and then kill it every 3 minutes. Only them I can get any suggestions out of nimsuggest.
Nowadays, VSCode nim plugin comes with useful setting: `nimssugestResetTimeout`.

**Comment by dom96** [comment]
Just to be clear: I'm not excusing nimsuggest here. But the way the VS Code extension launches nimsuggest is incredibly infuriating.

We should move to a Language Server that can manage nimsuggest instances in a sane manner.

**Comment by sinkingsugar** [comment]
btw for some reason it's not happening anymore lately!
(I pull `devel` quite often.)

**Comment by moigagoo** [comment]
JFYI setting nim.project in the workspace config helps:
```
{
  "nim.project": [
    "src/norm.nim"
  ]
}
```

**Comment by anurbol** [comment]
Hi, nimsuggest just ate 4 GB of RAM on my Windows 7 and constantly uses 100% of one of the cores

**Comment by mratsim** [comment]
@sinkingsugar might be related to https://github.com/nim-lang/Nim/pull/11849. That caused infinite loop in macros.

@anurbol, it's probably not a memory leak as CPU usage is high as well, probably loop that keeps burning CPUs and allocating, in a similar vein to https://github.com/nim-lang/Nim/pull/11849. Anyway that's what this issue is tracking.

**Comment by s0kil** [comment]
This issue still persists, I have noticed when working with HTTP server, if there is an issue in the code the `nimsuggest` process responsible for that file won't get killed even after the code has been fixed, I end up killing it manually.

**Comment by mratsim** [comment]
Yes, I have it quite often as well, and I'm not working with macros or a heavy generics codebase.

**Comment by shirleyquirk** [comment]
nimsuggest uses 100% cpu for me very regularly.  here's one way to replicate:
```nim
proc bar(k:static bool):SomeNumber = (when k: 3, else: 3.0)
```
running `nim check` on that file also causes an infinite loop, printing:
`Error: expression expected, but found ','`

i thought `nim check` was supposed to give up after some number of errors, no?

**Comment by arthurrasmusson** [comment]
I am also experiencing this issue in VSCode on MacOS. I believe I am on the latest version of the VSCode Nim plugin, VSCodium (open source vscode), and Nim.

**Comment by omentic** [comment]
I also see the `nimsuggest` and `nim check` processes eating up 100% of my CPU. 

![2022-05-14-152820](https://user-images.githubusercontent.com/35242550/168450134-c81dc43d-1c34-475e-b598-bf9fd2a2db6b.png)

**Comment by kanedaaaa** [comment]
Still an issue, wtf?? 

![Screenshot from 2023-11-04 14-35-48](https://github.com/nim-lang/Nim/assets/68785244/01d839cc-5217-441d-b1a9-81fd3a4d0281)
